File: app/src/main/java/io/aeyer/voidcore/ws/Envelope.java
package io.aeyer.voidcore.ws;

import com.fasterxml.jackson.databind.JsonNode;

/**
 * Wire envelope for every WS frame, in both directions. Per SPEC §4.2.
 *
 * <p>{@code seq} and {@code mac} are reserved for v2 per-message authentication
 * (ADR-018) and are accepted/ignored on v1. The dispatcher in {@link MessageDispatcher}
 * decodes {@link #payload} into a typed message variant; this record stays
 * untyped at the transport layer so #14's sealed protocol types can be
 * introduced without disturbing the wire pipe.
 */
public record Envelope(
        String id,
        String type,
        String protocol_version,
        long seq,
        String mac,
        JsonNode payload
) {
}

File: app/src/main/frontend/src/envelope.ts
import { type Envelope, PROTOCOL_VERSION } from "./types.js";

/** Build an outbound envelope. seq=0 / mac=null per SPEC §4.2 (v1). */
export function envelope<P>(type: string, payload: P, id: string | null = null): Envelope<P> {
  return {
    id: id ?? makeId(),
    type,
    protocol_version: PROTOCOL_VERSION,
    seq: 0,
    mac: null,
    payload,
  };
}

let counter = 0;
function makeId(): string {
  counter = (counter + 1) & 0xffffff;
  return `c-${Date.now().toString(36)}-${counter.toString(36)}`;
}
